### PR TITLE
Add ecmh go implementation

### PR DESCRIPTION
**Pull request by cpacia**
Spec here: https://github.com/tomasvdw/bips/blob/master/ecmh.mediawiki

### CODE REVIEW

File: bchec/ecmh.go [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+// Remove hashes the data onto the curve and subtracts the value
+// from the state. This function will execute regardless of whether
+// or not the passed data was previously added to the set. Hence,
+// adding a bunch of elements and then removing them all will not
+// result in a zero hash.
+func (ms *Multiset) Remove(data []byte) {
+	x, y := hashToPoint(ms.curve, data)
+	fx1, fy1 := ms.curve.bigAffineToField(x, y)
+	fy1 = fy1.Negate(1)
```
Comment (cpacia) [inline bchec/ecmh.go]: I don't know if using 1 for the magnitude here is always correct. It passes the test vectors but is this value always correct?

File: bchec/ecmh.go [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+	var x, y *big.Int
+	var err error
+	for {
+		h := sha256.Sum256(data)
+		n := make([]byte, 8)
+		binary.LittleEndian.PutUint64(n, i)
+		h2 := sha256.Sum256(append(n, h[:]...))
+
+		x = new(big.Int).SetBytes(h2[:])
+		y, err = decompressPoint(curve, x, false)
```
Comment (cpacia) [inline bchec/ecmh.go]: I _think_ `decompressPoint` is correct here but I'm not 100% sure. 

The spec says:
> If x is an element of the field and x3+7 is a quadratic residue, then P({d}) = (x, (x3+7)1/2)

File: bchec/ecmh.go [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+// hashToPoint hashes the passed data into a point on the curve. The x value
+// is sha256(n, sha256(data)) where n starts at zero. If the resulting x value
+// is not in the field or x^3+7 is not quadratic residue then n is incremented
+// and we try again. There is a 50% chance of success for any given iteration.
+func hashToPoint(curve *KoblitzCurve, data []byte) (*big.Int, *big.Int) {
+	i := uint64(0)
+	var x, y *big.Int
+	var err error
+	for {
```
Comment (tyler-smith) [inline bchec/ecmh.go]: Is it guaranteed that some that value of `i` will produce values that can `decompressPoint` without erroring? It seems like this is prone to getting stuck in an infinite loop.

Comment (cpacia) [reply bchec/ecmh.go]: It's a 50% chance it will not error afaik.

File: bchec/ecmh.go (hunk 1 above)
Comment (cpacia) [reply bchec/ecmh.go]: Created a multiset with millions of elements and removing each one still returned the correct hash. That would seem to suggest 1 is the correct value.

Comment (zquestz) [reply bchec/ecmh.go]: Yes, I would say this is correct.

**Pull request merged**